internal/compiler/transforms/compile/css-handler/prefix/utils.ts

```diff
@@ -0,0 +1,246 @@
+import {
+	CSSBlock,
+	CSSDeclaration,
+	cssBlock,
+	cssDeclaration,
+	cssIdentifier,
+} from "@internal/ast";
+import {UnknownObject} from "@internal/typescript-helpers";
+import {
+	CompilerPath,
+	EnterSignal,
+	ExitSignal,
+	createVisitor,
+	signals,
+} from "@internal/compiler";
+import {
+	VisitorStateEnter,
+	VisitorStateExit,
+} from "@internal/compiler/lib/VisitorState";
+import {Browser} from "@internal/browser-features/Browser";
+
+export function nodeHasPrefixedProperty(
+	node: CSSBlock,
+	property: string,
+	prefix: string,
+	rename: (propertyName: string) => string = (propertyName) => propertyName,
+): boolean {
+	if (node.value) {
+		return node.value.some((n) =>
+			n.type === "CSSDeclaration" && n.name === rename(`-${prefix}-${property}`)
+		);
+	}
+	return false;
+}
+
+export function nodeHasPrefixedPropertyValue(
+	node: CSSBlock,
+	property: string,
+	value: string,
+	prefix: string,
+	rename: (value: string) => string = (value) => value,
+): boolean {
+	if (node.value) {
+		return node.value.some((n) =>
+			n.type === "CSSDeclaration" &&
+			n.name === property &&
+			n.value.length === 1 &&
+			n.value[0]?.type === "CSSIdentifier" &&
+			n.value[0].value === rename(`-${prefix}-${value}`)
+		);
+	}
+	return false;
+}
+
+export function nodePropertyIndex(node: CSSBlock, property: string): number {
+	return node.value
+		? node.value.findIndex((n) =>
+				n.type === "CSSDeclaration" && n.name === property
+			)
+		: -1;
+}
+
+export function nodePropertyValueIndex(
+	node: CSSBlock,
+	property: string,
+	value: string,
+): number {
+	return node.value
+		? node.value.findIndex((n) =>
+				n.type === "CSSDeclaration" &&
+				n.name === property &&
+				n.value.length === 1 &&
+				n.value[0]?.type === "CSSIdentifier" &&
+				n.value[0].value === value
+			)
+		: -1;
+}
+export interface PrefixVisitor<State extends UnknownObject> {
+	name: string;
+	enter?: (
+		path: CompilerPath,
+		targets: Browser[],
+		state: VisitorStateEnter<State>,
+	) => EnterSignal;
+	exit?: (
+		path: CompilerPath,
+		targets: Browser[],
+		state: VisitorStateExit<State>,
+	) => ExitSignal;
+}
+
+export function createPrefixVisitor<State extends UnknownObject>(
+	visitor: PrefixVisitor<State>,
+) {
+	return {
+		name: `css-handler/prefix/${visitor.name}`,
+		enter: visitor.enter,
+		exit: visitor.exit,
+	};
+}
+
+export function wrapPrefixVisitor<State extends UnknownObject>(
+	visitor: PrefixVisitor<State>,
+	targets: Browser[],
+) {
+	return createVisitor<State>({
+		name: visitor.name,
+		enter: visitor.enter != null
+			? (path, state) => visitor.enter!(path, targets, state)
+			: undefined,
+		exit: visitor.exit != null
+			? (path, state) => visitor.exit!(path, targets, state)
+			: undefined,
+	});
```

Comment:
If we assign the properties to variables first then TypeScript wont lose the refinement.

```suggestion
    const {enter, exit} = visitor;
	return createVisitor<State>({
		name: visitor.name,
		enter: enter === undefined
			? undefined
			: (path, state) => enter(path, targets, state),
		exit: exit === undefined
			? undefined
			: (path, state) => exit(path, targets, state),
	});
```